src/MergeContext.js

```diff
@@ -318,13 +301,35 @@ class MergeContext {
         if (what === "postcondition")
             await this._setApprovalStatus(this._tagSha);
 
+        if (this._prInProgress()) {
+            this._log(what + " 'not in progress' failed");
+            return StepResult.Fail();
+        }
+
+        if (!this._prMergeable()) {
+            this._log(what + " 'mergeable' failed");
+            return StepResult.Fail();
+        }
+
+        if (what === "precondition") {
+            if (await this._stagingFailed()) {
+                this._log(what + " 'fresh tag with failed staging checks' failed'");
+                return StepResult.Fail();
+            }
+        }
+
         const commitStatus = await this._checkStatuses(this._prHeadSha());
-        /// status checks either failed, or pending and we don't know the time to rerun in
+        /// status checks either failed, or pending (except the approval check, having timeout)
         if (commitStatus === 'failure' || (commitStatus === 'pending' && !this._approval.grantedTimeout())) {
```

Comment:
I am not sure (because I cannot fully understand _checkStatuses() details), but it seems possible that commitStatus is `pending` because there is a real pending test _and_ a pending approval? In that case, the second condition will fail because this._approval.grantedTimeout() will be true.

If there is a bug here, then fixing it may be outside this PR scope. Your call.

Reply:
> but it seems possible that commitStatus is pending because there is a real pending test and a pending approval

Yes, it is possible we have 'pending approval' (e.g., 'fast track', grantedTimeout() is true), and some of the checks are 'pending'. This case will be processed similarly as if no other 'pending' checks exist: the PR processing will finish, enabling rerun in 'fast track' time left. I see no bug here.

Reply:
I will simplify the presentation a little by ignoring message validity and this._approval.granted() checks after the complex `if` statement.

To me, the code in question reads like this:

```
if (some test has failed)
    return failure;

if (some test is pending and it is not our fake test that waits for objections)
    return failure;

/* all real tests are successful, merge now or wait more for objections */
...
```

which would be OK. However, AFAICT, you are saying that the code is actually doing this:

```
if (some test failed)
    return failure;

if (some tests are pending but we are not waiting for objections)
    return failure;

/* XXX: some real tests may be still running here, but we ignore them */
/* merge now or wait for objections */
...
```

I do not know whether the XXX above leads to what you might call "real bugs", but it feels dangerously wrong at best.

Reply:
I tried to address your concerns in [PR17](https://github.com/measurement-factory/anubis/pull/17/files).